File: java/dsa/data-structures/stack-queue/remove-k-digits.java
class Solution {
    public String removeKdigits(String num, int k) {
        Deque<Character> dq = new ArrayDeque<>(); // @viz: init_stack
        
        for (char digit : num.toCharArray()) { // @viz: loop
            while (k > 0 && !dq.isEmpty() && dq.peekLast() > digit) { // @viz: while_check
                dq.removeLast(); // @viz: pop
                k--; // @viz: dec_k
            }
            dq.addLast(digit); // @viz: push
        }
        
        while (k > 0 && !dq.isEmpty()) { // @viz: trim_check
            dq.removeLast(); // @viz: pop_trim
            k--; // @viz: dec_k_trim
        }
        
        StringBuilder sb = new StringBuilder();
        boolean hasLeading = true;
        for (char d : dq) {
            if (hasLeading && d == '0') continue;
            hasLeading = false;
            sb.append(d);
        }
        
        return sb.length() == 0 ? "0" : sb.toString(); // @viz: finish
    }
}

File: javascript/dsa/data-structures/stack-queue/remove-k-digits.js
function removeKdigits(num, k) {
  const stack = []; // @viz: init_stack
  
  for (let digit of num) { // @viz: loop
    while (k > 0 && stack.length > 0 && stack[stack.length - 1] > digit) { // @viz: while_check
      stack.pop(); // @viz: pop
      k--; // @viz: dec_k
    }
    stack.push(digit); // @viz: push
  }
  
  while (k > 0) { // @viz: trim_check
    stack.pop(); // @viz: pop_trim
    k--; // @viz: dec_k_trim
  }
  
  let res = stack.join("").replace(/^0+/, ""); // @viz: join
  return res === "" ? "0" : res; // @viz: finish
}